lib/AST/ArgumentList.cpp

```diff
@@ -132,9 +132,10 @@ ArgumentList::createImplicit(ASTContext &ctx, ArrayRef<Argument> args,
                         firstTrailingClosureIndex, arena);
 }
 
-ArgumentList *ArgumentList::forImplicitSingle(ASTContext &ctx, Identifier label,
-                                              Expr *arg) {
-  return createImplicit(ctx, {Argument(SourceLoc(), label, arg)});
+ArgumentList *ArgumentList::forImplicitSingle(ASTContext &ctx, SourceLoc loc,
+                                              Identifier label, Expr *arg) {
+  return createImplicit(ctx, /*lParenLoc=*/loc, {Argument(loc, label, arg)},
+                        /*lParenLoc=*/loc);
```

Comment:
One potential concern with this is that it could result in out-of-order source ranges where the arg expression extends past the argument list, IMO it would make more sense to compute the paren locs from the start and end location of the expression.

An alternative fix here would be to change `KeyPathExpr::Component::for(Unresolved)Subscript` to use `ArgumentList::getStartLoc` instead of `ArgumentList::getLParenLoc`, I don't feel too strongly either way though.

Reply:
Why is this a concern for implicit expressions? Like hitting an assert in source range computation where the start is past the end?

Reply:
It's just a useful invariant that we ought to be able to enforce for the AST in general, though currently the ASTVerifier avoids checking implicit nodes

Reply:
> An alternative fix here would be to change `KeyPathExpr::Component::for(Unresolved)Subscript` to use `ArgumentList::getStartLoc` instead of `ArgumentList::getLParenLoc`

This makes more sense to me, thank you!